### PR TITLE
Add `--skip` flag to exclude tests

### LINKED ISSUE #1512: Add exclusion of the tests in CLI
### Which components does the task require to be changed?

snforge

### Description

We should add a regex which would make it able to exclude certain file/path/test name patterns

We should also investigate how other tools approach this

### PR DESCRIPTION
**Pull request by kkawula**
Closes #1512

## Introduced changes

-

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: docs/src/testing/running-tests.md [hunk 1]
```diff
@@ -71,6 +71,30 @@ Tests: 1 passed, 0 failed, 0 skipped, 0 ignored, other filtered out
 </details>
 <br>
 
+## Skipping tests
+
+You can use the `--skip` flag to run all tests _except_ those matching a specified filter pattern.
+This is useful for temporarily excluding problematic tests or focusing on a subset of tests by excluding others.
+
+<!-- { "package_name": "failing_example", "ignored_output": false } -->
```
Comment (franciszekjob) [inline docs/src/testing/running-tests.md]: It's default:
```suggestion
<!-- { "package_name": "failing_example" } -->
```

File: crates/snforge-scarb-plugin/src/attributes/test.rs [hunk 2]
```diff
@@ -83,8 +89,6 @@ fn test_internal(
     } else {
         Ok(formatdoc!(
             "
-            #[{config}]
-            {func_item}
```
Comment (franciszekjob) [inline crates/snforge-scarb-plugin/src/attributes/test.rs]: Why was this removed?

File: docs/src/appendix/snforge/test.md [hunk 3]
```diff
@@ -13,6 +13,10 @@ containing this filter.
 Will only run a test with a name exactly matching the test filter.
 Test filter must be a whole qualified test name e.g. `package_name::my_test` instead of just `my_test`.
 
+## `--skip <SKIP>`
+
+Skips any tests whose name contains the given SKIP string
```
Comment (franciszekjob) [inline docs/src/appendix/snforge/test.md]: nit:
```suggestion
Skips any tests whose name contains the given `SKIP` string
```

File: crates/forge/tests/e2e/running.rs
Comment (franciszekjob) [inline crates/forge/tests/e2e/running.rs:1]: Generally we use `[..]` for gas values in outputs, so let's maybe remove these specific values.

File: crates/forge/src/lib.rs [hunk 4]
```diff
@@ -152,6 +152,10 @@ pub struct TestArgs {
     #[arg(short, long)]
     exact: bool,
 
+    /// Skips any tests whose name contains the given SKIP string.
```
Comment (franciszekjob) [inline crates/forge/src/lib.rs:155]:
```suggestion
    /// Skips any tests whose name contains the given string.
```

Comment (kkawula) [reply crates/forge/src/lib.rs:155]: image

It makes sense in `snforge test --help`

File: crates/snforge-scarb-plugin/src/attributes/test.rs (hunk 2 above)
Comment (kkawula) [reply crates/snforge-scarb-plugin/src/attributes/test.rs]: @cptartur, why is it needed? XD bcs imo is unnecessary overhead

File: crates/forge/tests/e2e/running.rs [hunk 5]
```diff
@@ -223,6 +223,103 @@ fn with_exact_filter() {
...
+        "},
+    );
+}
+
+#[test]
+fn with_skip_filter_matching_multiple_test_name() {
+    let temp = setup_package("simple_package");
+
+    let output = test_runner(&temp)
+        .arg("--skip")
```
Comment (cptartur) [inline crates/forge/tests/e2e/running.rs:341]: Can `--skip` also take separated values somehow or do you have to prepend each with `--skip`?

File: crates/snforge-scarb-plugin/src/attributes/test.rs (hunk 2 above)
Comment (cptartur) [reply crates/snforge-scarb-plugin/src/attributes/test.rs]: Either you add these or not, they won't be compiled if they don't have an executable attribute.

File: docs/src/testing/running-tests.md [hunk 6]
```diff
@@ -71,6 +71,30 @@ Tests: 1 passed, 0 failed, 0 skipped, 0 ignored, other filtered out
 </details>
 <br>
 
+## Skipping tests
```
Comment (cptartur) [inline docs/src/testing/running-tests.md:74]: Maybe we should say that skipped tests are not compiled?

File: crates/snforge-scarb-plugin/src/attributes/test.rs (hunk 2 above)
Comment (cptartur) [reply crates/snforge-scarb-plugin/src/attributes/test.rs]: The only edge case I can come up with is that if the user has `SNFORGE_TEST_SKIP` set, then LS might behave weird because `#[test]` macro will start resolving to empty code for some tests.
It's super unlikely to happen but I'd rather keep this code then open ourselves to possible bugs.

File: crates/forge/src/run_tests/workspace.rs [hunk 7]
```diff
@@ -58,6 +58,11 @@ pub async fn run_for_workspace(args: TestArgs) -> Result<ExitStatus> {
         }
     }
 
+    if !args.skip.is_empty() {
+        let skip_patterns = args.skip.join(";");
+        set_forge_test_skip(skip_patterns);
```
Comment (cptartur) [inline crates/forge/src/run_tests/workspace.rs]: I'd move the logic for joining to `set_forge_test_skip`. Let's keep the "contract" for the `SNFORGE_TEST_SKIP` format in a relevant function.

File: crates/forge/tests/e2e/running.rs (hunk 5 above)
Comment (kkawula) [reply crates/forge/tests/e2e/running.rs:341]: That's how it works in cargo and clap, yes, we have prepend `--skip`

https://docs.rs/clap/latest/clap/_derive/_tutorial/index.html#options

or add value_delimiter: https://docs.rs/clap/latest/clap/struct.Arg.html#method.value_delimiter

File: docs/src/testing/running-tests.md (hunk 6 above)
Comment (kkawula) [reply docs/src/testing/running-tests.md:74]: It isn't true if we skip e.g. by module name

**Review by franciszekjob — APPROVED** [review]

File: crates/forge/tests/e2e/running.rs (hunk 5 above)
Comment (cptartur) [reply crates/forge/tests/e2e/running.rs:341]: I think its fine to prepend each

File: docs/src/testing/running-tests.md [hunk 8]
```diff
@@ -71,6 +71,30 @@ Tests: 1 passed, 0 failed, 0 skipped, 0 ignored, other filtered out
...
+
+You can use the `--skip` flag to run all tests _except_ those matching a specified filter pattern.
+This is useful for temporarily excluding problematic tests or focusing on a subset of tests by excluding others.
```
Comment (cptartur) [inline docs/src/testing/running-tests.md]:
```suggestion
You can use the `--skip` flag to exclude tests matching a specified filter pattern.
This is useful for temporarily disable problematic tests or focusing on a subset of tests by excluding others.
```

File: docs/src/testing/running-tests.md [hunk 9]
```diff
@@ -71,6 +71,30 @@ Tests: 1 passed, 0 failed, 0 skipped, 0 ignored, other filtered out
...
+
+<details>
+<summary>Output:</summary>
+
+```shell
+Collected 1 test(s) from failing_example package
+Running 0 test(s) from src/
+Running 1 test(s) from tests/
+[PASS] failing_example_tests::test_abc (l1_gas: ~0, l1_data_gas: ~0, l2_gas: ~40000)
+Tests: 1 passed, 0 failed, 0 skipped, 0 ignored, 0 filtered out
```
Comment (cptartur) [inline docs/src/testing/running-tests.md]: Why do it says that 0 packages were skipped?

File: docs/src/testing/running-tests.md [hunk 10]
```diff
@@ -71,6 +71,30 @@ Tests: 1 passed, 0 failed, 0 skipped, 0 ignored, other filtered out
...
 <br>
 
+## Skipping tests
+
+You can use the `--skip` flag to run all tests _except_ those matching a specified filter pattern.
+This is useful for temporarily excluding problematic tests or focusing on a subset of tests by excluding others.
+
+<!-- { "package_name": "failing_example"} -->
+```shell
+$ snforge test --skip test_failing --skip xyz
```
Comment (cptartur) [inline docs/src/testing/running-tests.md]: Isn't the name `xyz` a bit confusing to be used in an example?

File: crates/forge/tests/e2e/running.rs [hunk 11]
```diff
@@ -223,6 +223,103 @@ fn with_exact_filter() {
...
+        [PASS] simple_package_integrationtest::test_simple::test_simple2 [..]
+        [PASS] simple_package_integrationtest::ext_function_test::test_simple [..]
+        [PASS] simple_package_integrationtest::without_prefix::five [..]
+        [PASS] simple_package_integrationtest::ext_function_test::test_my_test [..]
+        [PASS] simple_package_integrationtest::test_simple::test_simple [..]
+        [PASS] simple_package_integrationtest::contract::call_and_invoke [..]
+        Running 2 test(s) from src/
+        [IGNORE] simple_package::tests::ignored_test
+        [PASS] simple_package::tests::test_fib [..]
+        Tests: 7 passed, 0 failed, 0 skipped, 2 ignored, 0 filtered out
```
Comment (cptartur) [inline crates/forge/tests/e2e/running.rs]: Should these be counted as ignored and not skipped? 
I remember there was some discussion about it already but not sure why we have decided to count it as such.

File: crates/forge/tests/e2e/running.rs
Comment (cptartur) [inline crates/forge/tests/e2e/running.rs:1]: If we can skip by module, we should have tests that explicitly use module syntax for skipping.

**Review by cptartur — COMMENTED** [review]
Code looks good, just needs some improvements to the docs and tests.

File: docs/src/testing/running-tests.md [hunk 12]
```diff
@@ -66,11 +66,35 @@ Collected 1 test(s) from hello_snforge package
...
-Tests: 1 passed, 0 failed, 0 skipped, 0 ignored, other filtered out
+Tests: 1 passed, 0 failed, 0 ignored, other filtered out
 ```
 </details>
 <br>
 
+## Skipping tests
+
+You can use the `--skip` flag to exclude tests matching a specified filter pattern.
+This is useful for temporarily disable problematic tests or focusing on a subset of tests by excluding others.
```
Comment (ddoktorski) [inline docs/src/testing/running-tests.md]: Can we mention that it is also possible to skip by module name and path? An example using module syntax might be helpful

File: CHANGELOG.md [hunk 13]
```diff
@@ -35,6 +35,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - ETH token is now pre-deployed in every test by default
 - Added `Token::ETH` to snforge_std
+- `--skip` flag to exclude tests whose name contains the given string
```
Comment (ddoktorski) [inline CHANGELOG.md]: nit:
```suggestion
- `--skip` flag to allow excluding any test whose name contains the provided string
```

File: CHANGELOG.md [hunk 14]
```diff
@@ -35,6 +35,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 
 #### Changed
```
Comment (ddoktorski) [inline CHANGELOG.md]: Let's add an entry here to mention the change in the `snforge test` output: that `skipped: ` summary has been removed and `--exit-first` now returns different output.

Comment (kkawula) [reply CHANGELOG.md]: Done

File: CHANGELOG.md (hunk 13 above)
Comment (kkawula) [reply CHANGELOG.md]: Done

File: docs/src/testing/running-tests.md (hunk 12 above)
Comment (kkawula) [reply docs/src/testing/running-tests.md]: Done

File: CHANGELOG.md [hunk 15]
```diff
@@ -9,6 +9,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Forge
 
+#### Added
+- `--skip` flag to allow excluding any test whose name contains the provided string
+
+#### Changed
+- Updated output format for `--exit-first` flag. Tests skipped due to preceding failures are no longer displayed in the summary. Alternative information is shown when applicable.
+
 #### Changed
```
Comment (ddoktorski) [inline CHANGELOG.md]: Double "Changed" section, but we can leave it for now and fix it in the release PR

**Review by ddoktorski — APPROVED** [review]
